services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateManager.java

```diff
@@ -0,0 +1,354 @@
+        limitations under the License.
+*/
+
+package com.futurewei.alcor.dataplane.utils;
+
+import com.futurewei.alcor.dataplane.service.GoalStateService;
+import com.futurewei.alcor.schema.*;
+import com.futurewei.alcor.schema.Port.PortState;
+import com.futurewei.alcor.web.entity.dataplane.InternalPortEntity;
+import com.futurewei.alcor.web.entity.dataplane.InternalSubnetEntity;
+import com.futurewei.alcor.web.entity.dataplane.NeighborInfo;
+import com.futurewei.alcor.web.entity.dataplane.NetworkConfiguration;
+import com.futurewei.alcor.web.entity.port.PortEntity;
+import com.futurewei.alcor.web.entity.vpc.VpcEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+
+import static com.futurewei.alcor.schema.Port.PortConfiguration.FixedIp;
+
+@Component
+public class GoalStateManager {
+  @Autowired private GoalStateService goalStateService;
+  private static final Logger LOG = LoggerFactory.getLogger(GoalStateManager.class);
+  private Map<String, Set<String>> portsInSameSubnetMap = new HashMap<>();
+  private Map<String, Set<NeighborInfo>> neighborInfoInSameSubenetMap = new HashMap<>();
+
+  public Map<String, Goalstate.GoalState> transformNorthToSouth(
+      // get the data
+      NetworkConfiguration networkConfiguration) throws RuntimeException {
+    InternalPortEntity[] portStatesArr =
+        networkConfiguration.getPortEntities().toArray(new InternalPortEntity[0]);
+    InternalSubnetEntity[] subnetArr =
+        networkConfiguration.getSubnets().toArray(new InternalSubnetEntity[0]);
+    com.futurewei.alcor.web.entity.vpc.VpcEntity[] vpcArr =
+        networkConfiguration.getVpcs().toArray(new com.futurewei.alcor.web.entity.vpc.VpcEntity[0]);
+
+    // TODO need to refactor subnet and vpc part when logic is
+    //  clear and integration done
+    Map<String, List<InternalPortEntity>> mapGroupedByHostIp = new HashMap();
+    Map<String, InternalSubnetEntity> subnetMap = new HashMap<>();
+    Map<String, InternalPortEntity> portMap = new HashMap<>();
+    Map<String, com.futurewei.alcor.web.entity.vpc.VpcEntity> vpcMap = new HashMap<>();
+    // construct map from list
+    for (InternalSubnetEntity s : subnetArr) {
+      subnetMap.put(s.getId(), s);
+    }
+
+    for (VpcEntity vpc : vpcArr) {
+      vpcMap.put(vpc.getId(), vpc);
+    }
+    // group nb msg by ip
+    int portCounter = 0;
+    for (InternalPortEntity portEntityNB : portStatesArr) {
+      portMap.put(portEntityNB.getId(), portEntityNB);
+      String bindingHostIP = portEntityNB.getBindingHostIP();
+      InternalPortEntity currentPortEntity = portStatesArr[portCounter];
+      if (!mapGroupedByHostIp.containsKey(bindingHostIP)) {
+        List<InternalPortEntity> portStates = new ArrayList<>();
+        fillSubnetAndVpcToPort(subnetMap, vpcMap, currentPortEntity, portStates);
```

Comment:
Line 76-78 is same as Line 81-83. Could be consolidated.

Reply:
fixed in 7958f8b